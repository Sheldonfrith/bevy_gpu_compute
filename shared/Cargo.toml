[package]
name = "shared"
version = "0.1.0"
edition = "2024"

[dependencies]
bytemuck = { version= "1.21.0", features=["derive"]}
# spl-pod = "0.5.0"
paste = "1.0.15"
# proc-macro2 = "1.0.92"
# quote = "1.0.38"
[features]
f16 = []  # Enable unstable f16 data type support (16 bit floating point numbers), which can improve gpu compute performance if used

[dev-dependencies]
pretty_assertions = "1.4.1"
