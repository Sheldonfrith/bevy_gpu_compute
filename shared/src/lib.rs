#![feature(f16)]

pub mod compose_wgsl_module;
pub mod custom_type_name;
pub mod wgsl_components;
pub mod wgsl_in_rust_helpers;
pub mod wgsl_wgpu_binding;
