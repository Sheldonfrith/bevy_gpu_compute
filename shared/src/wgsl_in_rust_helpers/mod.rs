pub mod matrices;
pub mod vectors;

pub use matrices::*;
pub use vectors::*;

pub struct WgslGlobalId {
    pub x: u32,
    pub y: u32,
    pub z: u32,
}

/// This actually refers to wgpu "Uniforms", but most people don't know what those are, so we call them "configs" instead.
pub struct WgslConfigInput {}
impl WgslConfigInput {
    pub fn get<T>() -> T {
        unimplemented!()
    }
}

/// These method are named "vec" because per this library API you input your data as variable-sized vectors. But keep in mind that on the actual GPU these are all fixed-length arrays.
pub struct WgslVecInput {}
impl WgslVecInput {
    pub fn vec_len<T>() -> u32 {
        unimplemented!()
    }
    pub fn vec_val<T>(index: u32) -> T {
        unimplemented!()
    }
}
/**
 * All outputs are arrays/vectors.
 * No "get" type methods are implemented, sinc GPU operations are massively parallel, and you should not be READING from your outputs since you will have no way of knowing if another thread has already touched a certain output or not handled it yet. //todo: (need to add a link to an article explaining this)
 */
pub struct WgslOutput {}
impl WgslOutput {
    /// Using "array_set" is generally more performant, use this only if you can't determine beforehand the number of outputs you will be producing.
    /// Using this will generate an atomic counter for the specific output. If your ACTUAL output length is always very close to your maximum output length for this specific output, consider using "array_set" and manually removing the trailing empty values instead.
    /// However if your ACTUAL output length could be significantly smaller than your set maximum output length, then using "vec_push" is more performant. It was created for those situations. For example: collision detection where the actual number of collisions is often far less than the theoretical maximum.
    pub fn push<T>(val: T) {
        unimplemented!()
    }
    pub fn set<T>(index: u32, val: T) {
        unimplemented!()
    }
    /// returns the user-input maximum number of elements that can be stored in the output for this specific type.
    pub fn len<T>() -> u32 {
        unimplemented!()
    }
}
