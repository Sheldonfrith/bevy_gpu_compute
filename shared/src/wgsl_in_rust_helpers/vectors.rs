macro_rules! impl_vector {
    ($name:ident, $type:ty, $($field:ident, $index:expr),+) => {
        #[derive(Debug, Clone,Copy)]
        pub struct $name {
            $(pub $field: $type,)+
            _force_constructor: ()
        }

        impl $name {
            pub fn new($($field: $type),+) -> Self {
                Self {
                    $($field,)+
                    _force_constructor: ()
                }
            }

            // Generate setters for coordinate access (x,y,z,w)
            $(
                paste::paste! {
                    pub fn [<set_ $field>](&mut self, value: $type) {
                        self.$field = value;
                    }
                }
            )+
        }

        impl std::ops::Index<i32> for $name {
            type Output = $type;

            fn index(&self, index: i32) -> &Self::Output {
                match index {
                    $($index => &self.$field,)+
                    _ => panic!("Index out of bounds"),
                }
            }
        }

        impl std::ops::IndexMut<i32> for $name {
            fn index_mut(&mut self, index: i32) -> &mut Self::Output {
                match index {
                    $($index => &mut self.$field,)+
                    _ => panic!("Index out of bounds"),
                }
            }
        }
    }
}

macro_rules! define_vector_types {
    ($type:ty, $suffix:ident) => {
        paste::paste! {
            impl_vector!([<Vec2 $suffix>], $type, x, 0, y, 1);
            impl_vector!([<Vec3 $suffix>], $type, x, 0, y, 1, z, 2);
            impl_vector!([<Vec4 $suffix>], $type, x, 0, y, 1, z, 2, w, 3);


        }
    };
}

// Generate concrete types for all scalar types
define_vector_types!(bool, Bool);
define_vector_types!(u32, U32);
define_vector_types!(i32, I32);
define_vector_types!(f32, F32);
define_vector_types!(f16, F16);

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_vec4_f32_creation_and_access() {
        let mut vec4 = Vec4F32::new(1.0, 2.0, 3.0, 4.0);
        // Test direct field access
        assert_eq!(vec4.x, 1.0);
        assert_eq!(vec4.y, 2.0);
        assert_eq!(vec4.z, 3.0);
        assert_eq!(vec4.w, 4.0);

        // Test index access
        assert_eq!(vec4[0], 1.0);
        assert_eq!(vec4[1], 2.0);
        assert_eq!(vec4[2], 3.0);
        assert_eq!(vec4[3], 4.0);

        // Test setters
        vec4.x = 5.0;
        vec4[2] = 7.0;

        assert_eq!(vec4.x, 5.0);
        assert_eq!(vec4.y, 2.0);
        assert_eq!(vec4.z, 7.0);
        assert_eq!(vec4.w, 4.0);
    }

    #[test]
    #[should_panic(expected = "Index out of bounds")]
    fn test_vec4_f32_invalid_index() {
        let vec4 = Vec4F32::new(1.0, 2.0, 3.0, 4.0);
        let _value = vec4[4]; // Should panic
    }
}
