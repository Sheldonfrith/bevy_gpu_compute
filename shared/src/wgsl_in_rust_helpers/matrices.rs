use crate::wgsl_in_rust_helpers::vectors::*;
macro_rules! impl_matrix {
    ($name:ident, $vec_type:ty, $($field:ident, $index:expr),+) => {
        #[derive(Debug, Clone)]
        pub struct $name {
            $(pub $field: $vec_type,)+
            _force_constructor: ()
        }

        impl $name {
            pub fn new($($field: $vec_type),+) -> Self {
                Self {
                    $($field,)+
                    _force_constructor: ()
                }
            }

            // Generate setters for row access (x,y,z,w)
            $(
                paste::paste! {
                    pub fn [<set_ $field>](&mut self, value: $vec_type) {
                        self.$field = value;
                    }
                }
            )+
        }

        impl std::ops::Index<i32> for $name {
            type Output = $vec_type;

            fn index(&self, index: i32) -> &Self::Output {
                match index {
                    $($index => &self.$field,)+
                    _ => panic!("Index out of bounds"),
                }
            }
        }

        impl std::ops::IndexMut<i32> for $name {
            fn index_mut(&mut self, index: i32) -> &mut Self::Output {
                match index {
                    $($index => &mut self.$field,)+
                    _ => panic!("Index out of bounds"),
                }
            }
        }
    }
}

macro_rules! define_matrix_types {
    ($scalar_type:ty, $suffix:ident) => {
        paste::paste! {
            // Mat2xN types
            impl_matrix!([<Mat2x2 $suffix>], [<Vec2 $suffix>], x, 0, y, 1);
            impl_matrix!([<Mat2x3 $suffix>], [<Vec3 $suffix>], x, 0, y, 1);
            impl_matrix!([<Mat2x4 $suffix>], [<Vec4 $suffix>], x, 0, y, 1);

            // Mat3xN types
            impl_matrix!([<Mat3x2 $suffix>], [<Vec2 $suffix>], x, 0, y, 1, z, 2);
            impl_matrix!([<Mat3x3 $suffix>], [<Vec3 $suffix>], x, 0, y, 1, z, 2);
            impl_matrix!([<Mat3x4 $suffix>], [<Vec4 $suffix>], x, 0, y, 1, z, 2);

            // Mat4xN types
            impl_matrix!([<Mat4x2 $suffix>], [<Vec2 $suffix>], x, 0, y, 1, z, 2, w, 3);
            impl_matrix!([<Mat4x3 $suffix>], [<Vec3 $suffix>], x, 0, y, 1, z, 2, w, 3);
            impl_matrix!([<Mat4x4 $suffix>], [<Vec4 $suffix>], x, 0, y, 1, z, 2, w, 3);
        }
    };
}

// Generate concrete types for numeric types (matrices don't make sense for booleans)
define_matrix_types!(u32, U32);
define_matrix_types!(i32, I32);
define_matrix_types!(f32, F32);
define_matrix_types!(f16, F16);
define_matrix_types!(bool, Bool);

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_mat3x4_f32_creation_and_access() {
        let vec1 = Vec4F32::new(1.0, 2.0, 3.0, 4.0);
        let vec2 = Vec4F32::new(5.0, 6.0, 7.0, 8.0);
        let vec3 = Vec4F32::new(9.0, 10.0, 11.0, 12.0);

        let mut mat = Mat3x4F32::new(vec1, vec2, vec3);

        // Test direct field access
        assert_eq!(mat.x.x, 1.0);
        assert_eq!(mat.y.y, 6.0);
        assert_eq!(mat.z.z, 11.0);

        // Test index access
        assert_eq!(mat[0][0], 1.0);
        assert_eq!(mat[1][1], 6.0);
        assert_eq!(mat[2][2], 11.0);

        // Test setters
        let new_vec = Vec4F32::new(13.0, 14.0, 15.0, 16.0);
        mat.set_x(new_vec);

        assert_eq!(mat.x.w, 16.0);
        assert_eq!(mat[0][3], 16.0);
    }

    #[test]
    #[should_panic(expected = "Index out of bounds")]
    fn test_mat3x4_f32_invalid_index() {
        let vec1 = Vec4F32::new(1.0, 2.0, 3.0, 4.0);
        let vec2 = Vec4F32::new(5.0, 6.0, 7.0, 8.0);
        let vec3 = Vec4F32::new(9.0, 10.0, 11.0, 12.0);

        let mat = Mat3x4F32::new(vec1, vec2, vec3);
        let _value = mat[3]; // Should panic
    }
    #[test]
    fn test_mat4x2_bool_access_and_mutation() {
        let vec1 = Vec2Bool::new(true, false);
        let vec2 = Vec2Bool::new(false, true);
        let vec3 = Vec2Bool::new(true, true);
        let vec4 = Vec2Bool::new(false, false);

        let mut mat = Mat4x2Bool::new(vec1, vec2, vec3, vec4);

        // Test direct field access
        assert_eq!(mat.x.x, true);
        assert_eq!(mat.y.y, true);
        assert_eq!(mat.z.x, true);
        assert_eq!(mat.w.y, false);

        // Test index access
        assert_eq!(mat[0][0], true);
        assert_eq!(mat[1][1], true);
        assert_eq!(mat[2][0], true);
        assert_eq!(mat[3][1], false);

        // Test setters
        let new_vec = Vec2Bool::new(false, true);
        mat.set_x(new_vec);

        assert_eq!(mat.x.y, true);
        assert_eq!(mat[0][1], true);
    }
}
