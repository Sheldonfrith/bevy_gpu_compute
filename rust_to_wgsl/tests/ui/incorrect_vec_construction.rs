use shared::wgsl_in_rust_helpers::*;
const MY_CONST: Vec3Bool = Vec3Bool {
    x: true,
    y: false,
    z: true,
};
