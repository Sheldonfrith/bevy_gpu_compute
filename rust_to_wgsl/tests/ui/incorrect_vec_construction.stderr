error[E0601]: `main` function not found in crate `$CRATE`
 --> tests/ui/incorrect_vec_construction.rs:6:3
  |
6 | };
  |   ^ consider adding a `main` function to `$DIR/tests/ui/incorrect_vec_construction.rs`

error: cannot construct `shared::wgsl_in_rust_helpers::Vec3Bool` with struct literal syntax due to private fields
 --> tests/ui/incorrect_vec_construction.rs:2:28
  |
2 | const MY_CONST: Vec3Bool = Vec3Bool {
  |                            ^^^^^^^^
  |
  = note: ...and other private field `_force_constructor` that was not provided
help: you might have meant to use the `new` associated function
  |
2 | const MY_CONST: Vec3Bool = Vec3Bool::new(_, _, _);
  |                                    ~~~~~~~~~~~~~~
